### PR TITLE
Upgrade to Django 4.2 LTS

### PR DESCRIPTION
**Pull request by john-westcott-iv**
##### SUMMARY

Upgrade for Django LTS

Intended to replace #13892

This PR is not updating psycopg3, that will be performed in a different PR. 
Also had to upgrade some libraries for compatibility with the new Django:
  * djangorestframework
  * django-oauth-toolkit

Restframework and so far look not too bad but django-oauth-toolkit caused issues.
After reverting from django-oauth-toolkit 2.x down to 1.7 its fine. 2.x had breaking changes that definitely broke us. In a future effort we should attempt a specific upgrade of django-oauth-toolkit and work through its issues.

To Do:

- [x] Get CI checks green
- [x] Run all collection integration tests
- [x] Fix deprecation errors from tests
- [x] Test in different configs w and w/o FIPS mode

##### ISSUE TYPE

 - New or Enhanced Feature
 
##### COMPONENT NAME

 - Other

##### AWX VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: requirements/requirements.txt [hunk 1]
```diff
@@ -266,10 +274,10 @@ prometheus-client==0.15.0
     # via -r /awx_devel/requirements/requirements.in
 psutil==5.9.4
     # via -r /awx_devel/requirements/requirements.in
-psycopg==3.1.4
-    # via -r /awx_devel/requirements/requirements.in
 psycopg2==2.9.5
     # via -r /awx_devel/requirements/requirements.in
+psycopg2-binary==2.9.6
+    # via aiopg
```
Comment (AlanCoding) [inline requirements/requirements.txt]: I'm not savvy enough with the history to understand why we didn't use `psycopg2-binary` before. Maybe check with productization?

Comment (john-westcott-iv) [reply requirements/requirements.txt]: This is actually from a cherry picked commit from #13947. @TheRealHaoLiu, FYI

File: awx/main/migrations/0113_v370_event_bigint.py [hunk 2]
```diff
@@ -22,11 +22,8 @@ def migrate_event_data(apps, schema_editor):
             # recreate counter for the new table's primary key to
             # start where the *old* table left off (we have to do this because the
             # counter changed from an int to a bigint)
-            cursor.execute(f'DROP SEQUENCE IF EXISTS "{tblname}_id_seq" CASCADE;')
-            cursor.execute(f'CREATE SEQUENCE "{tblname}_id_seq";')
-            cursor.execute(f'ALTER TABLE "{tblname}" ALTER COLUMN "id" ' f"SET DEFAULT nextval('{tblname}_id_seq');")
-            cursor.execute(f"SELECT setval('{tblname}_id_seq', (SELECT MAX(id) FROM _old_{tblname}), true);")
-            cursor.execute(f'DROP TABLE _old_{tblname};')
+            cursor.execute(f'CREATE SEQUENCE IF NOT EXISTS "{tblname}_id_seq";')
+            cursor.execute(f"SELECT setval('{tblname}_id_seq', COALESCE((SELECT MAX(id)+1 FROM _old_{tblname}), 1), false);")
```
Comment (AlanCoding) [inline awx/main/migrations/0113_v370_event_bigint.py]: Could you give a little bit more elaboration on what happened with the removed lines here?

File: awx/main/migrations/0183_pre_django_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            ),
+        ),
+        migrations.AddField(
+            model_name='oauth2application',
+            name='algorithm',
+            field=models.CharField(
+                blank=True, choices=[('', 'No OIDC support'), ('RS256', 'RSA with SHA-2 256'), ('HS256', 'HMAC with SHA-2 256')], default='', max_length=5
+            ),
+        ),
+    ]
```
Comment (AlanCoding) [inline awx/main/migrations/0183_pre_django_upgrade.py:30]: For these 2 migration files, it might be helpful to share the `awx-manage sqlmigrate` output. This will help quickly clear up which steps are and are not making a schema change. I mostly expect to see non-schema changes as a result of a Django upgrade, as its own internal accounting changes.

File: awx/main/migrations/0113_v370_event_bigint.py (hunk 2 above)
Comment (john-westcott-iv) [reply awx/main/migrations/0113_v370_event_bigint.py:26]: Thanks for calling this out. I'm actually going to revert this file as it was more of a problem with the psycopg3 upgrade which is no longer a thing in this version of the PR.

Comment (john-westcott-iv) [reply awx/main/migrations/0113_v370_event_bigint.py]: I retract my previous statement about this being psycopg3. This was from the Django upgrade. When you try to run this in Django 4.2 with original lines you get this stack:
```
tools_awx_1 |   Applying main.0113_v370_event_bigint...Traceback (most recent call last):
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 87, in _execute
tools_awx_1 |     return self.cursor.execute(sql)
tools_awx_1 | psycopg2.errors.DependentObjectsStillExist: cannot drop sequence main_jobevent_id_seq because column id of table _old_main_jobevent requires it
tools_awx_1 | HINT:  You can drop column id of table _old_main_jobevent instead.
tools_awx_1 | 
tools_awx_1 | 
tools_awx_1 | The above exception was the direct cause of the following exception:
tools_awx_1 | 
tools_awx_1 | Traceback (most recent call last):
tools_awx_1 |   File "/usr/local/bin/awx-manage", line 9, in <module>
tools_awx_1 |     load_entry_point('awx', 'console_scripts', 'awx-manage')()
tools_awx_1 |   File "/awx_devel/awx/__init__.py", line 200, in manage
tools_awx_1 |     execute_from_command_line(sys.argv)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
tools_awx_1 |     utility.execute()
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
tools_awx_1 |     self.fetch_command(subcommand).run_from_argv(self.argv)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
tools_awx_1 |     self.execute(*args, **cmd_options)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/base.py", line 458, in execute
tools_awx_1 |     output = self.handle(*args, **options)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
tools_awx_1 |     res = handle_func(*args, **kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
tools_awx_1 |     post_migrate_state = executor.migrate(
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
tools_awx_1 |     state = self._migrate_all_forwards(
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
tools_awx_1 |     state = self.apply_migration(
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
tools_awx_1 |     state = migration.apply(state, schema_editor)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
tools_awx_1 |     operation.database_forwards(
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
tools_awx_1 |     self.code(from_state.apps, schema_editor)
tools_awx_1 |   File "/awx_devel/awx/main/migrations/0113_v370_event_bigint.py", line 25, in migrate_event_data
tools_awx_1 |     cursor.execute(f'DROP SEQUENCE IF EXISTS "{tblname}_id_seq" CASCADE;')
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 102, in execute
tools_awx_1 |     return super().execute(sql, params)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
tools_awx_1 |     return self._execute_with_wrappers(
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
tools_awx_1 |     return executor(sql, params, many, context)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
tools_awx_1 |     return self.cursor.execute(sql, params)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
tools_awx_1 |     raise dj_exc_value.with_traceback(traceback) from exc_value
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/utils.py", line 87, in _execute
tools_awx_1 |     return self.cursor.execute(sql)
tools_awx_1 | django.db.utils.InternalError: cannot drop sequence main_jobevent_id_seq because column id of table _old_main_jobevent requires it
tools_awx_1 | HINT:  You can drop column id of table _old_main_jobevent instead.
```

In `devel` (with Django 3.2) we see these are only NOTICE/DETAILS and not an ERROR:
```
tools_postgres_1 | 2023-05-11 15:14:28.033 UTC [86] NOTICE:  drop cascades to 2 other objects
tools_postgres_1 | 2023-05-11 15:14:28.033 UTC [86] DETAIL:  drop cascades to default value for column id of table _old_main_jobevent
tools_postgres_1 | 	drop cascades to default value for column id of table main_jobevent
tools_postgres_1 | 2023-05-11 15:14:28.072 UTC [86] NOTICE:  drop cascades to 2 other objects
tools_postgres_1 | 2023-05-11 15:14:28.072 UTC [86] DETAIL:  drop cascades to default value for column id of table _old_main_inventoryupdateevent
tools_postgres_1 | 	drop cascades to default value for column id of table main_inventoryupdateevent
tools_postgres_1 | 2023-05-11 15:14:28.125 UTC [86] NOTICE:  drop cascades to 2 other objects
tools_postgres_1 | 2023-05-11 15:14:28.125 UTC [86] DETAIL:  drop cascades to default value for column id of table _old_main_projectupdateevent
tools_postgres_1 | 	drop cascades to default value for column id of table main_projectupdateevent
tools_postgres_1 | 2023-05-11 15:14:28.190 UTC [86] NOTICE:  drop cascades to 2 other objects
tools_postgres_1 | 2023-05-11 15:14:28.190 UTC [86] DETAIL:  drop cascades to default value for column id of table _old_main_adhoccommandevent
tools_postgres_1 | 	drop cascades to default value for column id of table main_adhoccommandevent
tools_postgres_1 | 2023-05-11 15:14:28.238 UTC [86] NOTICE:  drop cascades to 2 other objects
tools_postgres_1 | 2023-05-11 15:14:28.238 UTC [86] DETAIL:  drop cascades to default value for column id of table _old_main_systemjobevent
tools_postgres_1 | 	drop cascades to default value for column id of table main_systemjobevent
```

Its unclear why the Django upgrade changes this from Notice to Error but this is why we had to alter these statements. 

However, since no one should be migrating from a version this old directly to this new version we can simplify this area greatly by just doing an alter table. Here is some details from psql showing that we can alter an int id to a bigint id and the sequence will also update:
```
awx=# CREATE TABLE test_table (id int GENERATED ALWAYS AS IDENTITY, value varchar);
CREATE TABLE
awx=# \d test_table_id_seq
                 Sequence "public.test_table_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Sequence for identity column: public.test_table.id
awx=# ALTER TABLE test_table ALTER COLUMN id TYPE bigint USING id::bigint;
ALTER TABLE
awx=# \d test_table_id_seq
                     Sequence "public.test_table_id_seq"
  Type  | Start | Minimum |       Maximum       | Increment | Cycles? | Cache 
--------+-------+---------+---------------------+-----------+---------+-------
 bigint |     1 |       1 | 9223372036854775807 |         1 | no      |     1
Sequence for identity column: public.test_table.id
```

Basically just altering the tables ID column to a bigint changes the sequence for us. Technically I think this would work for an upgrade scenario as well but since we truly only care about new table creation we can greatly simplify this to get around the error produced by Django 4.2/postgres.

File: awx/main/migrations/0183_pre_django_upgrade.py (hunk 3 above)
Comment (john-westcott-iv) [reply awx/main/migrations/0183_pre_django_upgrade.py:30]: 0183 is an oauth-toolkit migration and has legitimate output:
```
BEGIN;
--
-- Add field id_token to oauth2accesstoken
--
ALTER TABLE "main_oauth2accesstoken" ADD COLUMN "id_token_id" bigint NULL UNIQUE CONSTRAINT "main_oauth2accesstok_id_token_id_8a01d0b0_fk_oauth2_pr" REFERENCES "oauth2_provider_idtoken"("id") DEFERRABLE INITIALLY DEFERRED; SET CONSTRAINTS "main_oauth2accesstok_id_token_id_8a01d0b0_fk_oauth2_pr" IMMEDIATE;
--
-- Add field algorithm to oauth2application
--
ALTER TABLE "main_oauth2application" ADD COLUMN "algorithm" varchar(5) DEFAULT '' NOT NULL;
ALTER TABLE "main_oauth2application" ALTER COLUMN "algorithm" DROP DEFAULT;
COMMIT;
```

0184 is from the reindexing and, like you mentioned, all no-ops:
```
BEGIN;
--
-- Rename unnamed index for ('ad_hoc_command', 'job_created', 'uuid') on adhoccommandevent to main_adhocc_ad_hoc__1e4d24_idx
--
-- (no-op)
--
-- Rename unnamed index for ('ad_hoc_command', 'job_created', 'event') on adhoccommandevent to main_adhocc_ad_hoc__e72142_idx
--
-- (no-op)
--
-- Rename unnamed index for ('ad_hoc_command', 'job_created', 'counter') on adhoccommandevent to main_adhocc_ad_hoc__a57777_idx
--
-- (no-op)
--
-- Rename unnamed index for ('inventory_update', 'job_created', 'uuid') on inventoryupdateevent to main_invent_invento_f72b21_idx
--
-- (no-op)
--
-- Rename unnamed index for ('inventory_update', 'job_created', 'counter') on inventoryupdateevent to main_invent_invento_364dcb_idx
--
-- (no-op)
--
-- Rename unnamed index for ('job', 'job_created', 'parent_uuid') on jobevent to main_jobeve_job_id_40a56d_idx
--
-- (no-op)
--
-- Rename unnamed index for ('job', 'job_created', 'uuid') on jobevent to main_jobeve_job_id_3c4a4a_idx
--
-- (no-op)
--
-- Rename unnamed index for ('job', 'job_created', 'counter') on jobevent to main_jobeve_job_id_51c382_idx
--
-- (no-op)
--
-- Rename unnamed index for ('job', 'job_created', 'event') on jobevent to main_jobeve_job_id_0ddc6b_idx
--
-- (no-op)
--
-- Rename unnamed index for ('project_update', 'job_created', 'uuid') on projectupdateevent to main_projec_project_449bbd_idx
--
-- (no-op)
--
-- Rename unnamed index for ('project_update', 'job_created', 'counter') on projectupdateevent to main_projec_project_69559a_idx
--
-- (no-op)
--
-- Rename unnamed index for ('project_update', 'job_created', 'event') on projectupdateevent to main_projec_project_c44b7c_idx
--
-- (no-op)
--
-- Rename unnamed index for ('content_type', 'object_id') on role to main_rbac_r_content_979bdd_idx
--
-- (no-op)
--
-- Rename unnamed index for ('ancestor', 'content_type_id', 'role_field') on roleancestorentry to main_rbac_r_ancesto_b44606_idx
--
-- (no-op)
--
-- Rename unnamed index for ('ancestor', 'content_type_id', 'object_id') on roleancestorentry to main_rbac_r_ancesto_22b9f0_idx
--
-- (no-op)
--
-- Rename unnamed index for ('ancestor', 'descendent') on roleancestorentry to main_rbac_r_ancesto_c87b87_idx
--
-- (no-op)
--
-- Rename unnamed index for ('system_job', 'job_created', 'uuid') on systemjobevent to main_system_system__e39825_idx
--
-- (no-op)
--
-- Rename unnamed index for ('system_job', 'job_created', 'counter') on systemjobevent to main_system_system__73537a_idx
--
-- (no-op)
--
-- Alter field unified_job on activitystream
--
-- (no-op)
--
-- Alter field unified_job_template on activitystream
--
-- (no-op)
--
-- Alter field created_by on credential
--
-- (no-op)
--
-- Alter field modified_by on credential
--
-- (no-op)
--
-- Alter field created_by on credentialinputsource
--
-- (no-op)
--
-- Alter field modified_by on credentialinputsource
--
-- (no-op)
--
-- Alter field created_by on credentialtype
--
-- (no-op)
--
-- Alter field modified_by on credentialtype
--
-- (no-op)
--
-- Alter field created_by on custominventoryscript
--
-- (no-op)
--
-- Alter field modified_by on custominventoryscript
--
-- (no-op)
--
-- Alter field created_by on executionenvironment
--
-- (no-op)
--
-- Alter field credential on executionenvironment
--
-- (no-op)
--
-- Alter field modified_by on executionenvironment
--
-- (no-op)
--
-- Alter field organization on executionenvironment
--
-- (no-op)
--
-- Alter field created_by on group
--
-- (no-op)
--
-- Alter field modified_by on group
--
-- (no-op)
--
-- Alter field created_by on host
--
-- (no-op)
--
-- Alter field modified_by on host
--
-- (no-op)
--
-- Alter field smart_inventories on host
--
-- (no-op)
--
-- Alter field credential on instancegroup
--
-- (no-op)
--
-- Alter field created_by on inventory
--
-- (no-op)
--
-- Alter field modified_by on inventory
--
-- (no-op)
--
-- Alter field inventory on job
--
-- (no-op)
--
-- Alter field project on job
--
-- (no-op)
--
-- Alter field webhook_credential on job
--
-- (no-op)
--
-- Alter field credentials on joblaunchconfig
--
-- (no-op)
--
-- Alter field execution_environment on joblaunchconfig
--
-- (no-op)
--
-- Alter field instance_groups on joblaunchconfig
--
-- (no-op)
--
-- Alter field inventory on joblaunchconfig
--
-- (no-op)
--
-- Alter field labels on joblaunchconfig
--
-- (no-op)
--
-- Alter field inventory on jobtemplate
--
-- (no-op)
--
-- Alter field project on jobtemplate
--
-- (no-op)
--
-- Alter field webhook_credential on jobtemplate
--
-- (no-op)
--
-- Alter field created_by on label
--
-- (no-op)
--
-- Alter field modified_by on label
--
-- (no-op)
--
-- Alter field created_by on notificationtemplate
--
-- (no-op)
--
-- Alter field modified_by on notificationtemplate
--
-- (no-op)
--
-- Alter field user on oauth2accesstoken
--
-- (no-op)
--
-- Alter field user on oauth2application
--
-- (no-op)
--
-- Alter field created_by on organization
--
-- (no-op)
--
-- Alter field galaxy_credentials on organization
--
-- (no-op)
--
-- Alter field modified_by on organization
--
-- (no-op)
--
-- Alter field notification_templates_approvals on organization
--
-- (no-op)
--
-- Alter field notification_templates_error on organization
--
-- (no-op)
--
-- Alter field notification_templates_started on organization
--
-- (no-op)
--
-- Alter field notification_templates_success on organization
--
-- (no-op)
--
-- Alter field credential on project
--
-- (no-op)
--
-- Alter field signature_validation_credential on project
--
-- (no-op)
--
-- Alter field credential on projectupdate
--
-- (no-op)
--
-- Alter field created_by on schedule
--
-- (no-op)
--
-- Alter field credentials on schedule
--
-- (no-op)
--
-- Alter field execution_environment on schedule
--
-- (no-op)
--
-- Alter field inventory on schedule
--
-- (no-op)
--
-- Alter field labels on schedule
--
-- (no-op)
--
-- Alter field modified_by on schedule
--
-- (no-op)
--
-- Alter field created_by on team
--
-- (no-op)
--
-- Alter field modified_by on team
--
-- (no-op)
--
-- Alter field created_by on unifiedjob
--
-- (no-op)
--
-- Alter field credentials on unifiedjob
--
-- (no-op)
--
-- Alter field dependent_jobs on unifiedjob
--
-- (no-op)
--
-- Alter field execution_environment on unifiedjob
--
-- (no-op)
--
-- Alter field labels on unifiedjob
--
-- (no-op)
--
-- Alter field modified_by on unifiedjob
--
-- (no-op)
--
-- Alter field notifications on unifiedjob
--
-- (no-op)
--
-- Alter field organization on unifiedjob
--
-- (no-op)
--
-- Alter field polymorphic_ctype on unifiedjob
--
-- (no-op)
--
-- Alter field unified_job_template on unifiedjob
--
-- (no-op)
--
-- Alter field created_by on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field credentials on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field current_job on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field execution_environment on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field labels on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field last_job on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field modified_by on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field next_schedule on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field notification_templates_error on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field notification_templates_started on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field notification_templates_success on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field organization on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field polymorphic_ctype on unifiedjobtemplate
--
-- (no-op)
--
-- Alter field approved_or_denied_by on workflowapproval
--
-- (no-op)
--
-- Alter field inventory on workflowjob
--
-- (no-op)
--
-- Alter field webhook_credential on workflowjob
--
-- (no-op)
--
-- Alter field always_nodes on workflowjobnode
--
-- (no-op)
--
-- Alter field credentials on workflowjobnode
--
-- (no-op)
--
-- Alter field execution_environment on workflowjobnode
--
-- (no-op)
--
-- Alter field failure_nodes on workflowjobnode
--
-- (no-op)
--
-- Alter field inventory on workflowjobnode
--
-- (no-op)
--
-- Alter field labels on workflowjobnode
--
-- (no-op)
--
-- Alter field success_nodes on workflowjobnode
--
-- (no-op)
--
-- Alter field unified_job_template on workflowjobnode
--
-- (no-op)
--
-- Alter field inventory on workflowjobtemplate
--
-- (no-op)
--
-- Alter field notification_templates_approvals on workflowjobtemplate
--
-- (no-op)
--
-- Alter field webhook_credential on workflowjobtemplate
--
-- (no-op)
--
-- Alter field always_nodes on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field credentials on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field execution_environment on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field failure_nodes on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field inventory on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field labels on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field success_nodes on workflowjobtemplatenode
--
-- (no-op)
--
-- Alter field unified_job_template on workflowjobtemplatenode
--
-- (no-op)
COMMIT;
```

**Review by AlanCoding — APPROVED** [review]
Looks good to me, answered all my questions. Still need some QE eyes on it but I don't have any remaining open items.

**Review by akus062381 — APPROVED** [review]

**Review by masbahnana — APPROVED** [review]

File: requirements/requirements.in [hunk 4]
```diff
@@ -10,20 +10,20 @@ cryptography>=39.0.1 ## https://github.com/ansible/awx/security/dependabot/90
...
 distro
-django==3.2.16  # see UPGRADE BLOCKERs https://github.com/ansible/awx/security/dependabot/67
+django==4.2  # see UPGRADE BLOCKERs
 django-auth-ldap
 django-cors-headers
 django-crum
 django-extensions
 django-guid==3.2.1
-django-oauth-toolkit==1.4.1
+django-oauth-toolkit<2.0.0      # Version 2.0.0 has breaking changes that will need to be worked out before upgrading
```
Comment (jjwatt) [inline requirements/requirements.in:19]: A little hung up on this one, so I'll just comment/ask. This version specifier makes me a little nervous cus it's got no floor. So like:
```
~/git/github.com/ansible/jjwatt-awx devel*
awxvenv ❯ pip index versions django-oauth-toolkit 2> /dev/null  
django-oauth-toolkit (2.2.0)
Available versions: 2.2.0, 2.1.0, 2.0.0, 1.7.1, 1.7.0, 1.6.3, 1.6.2, 1.6.1, 1.6.0, 1.5.0, 1.4.1, 1.4.0, 1.3.3, 1.3.2, 1.3.0, 1.2.0, 1.1.3, 1.1.2, 1.1.1, 1.1.0, 1.0.0, 0.12.0, 0.11.0, 0.10.0, 0.9.0, 0.8.2, 0.8.1, 0.8.0, 0.7.2, 0.7.1, 0.7.0, 0.6.1, 0.6.0, 0.5.0, 0.4.1, 0.4.0, 0.3.2, 0.3.1, 0.3.0, 0.2.1, 0.2.0, 0.1.0
```
So, incompat with anything `2.0.0` or greater excludes the top 3 vers in the list, but there's a whole lot of others below. Is that alright?

**Review by jjwatt — CHANGES_REQUESTED** [review]
Just a quick question about that version specifier for `django-oauth-toolkit`

**Review by jjwatt — APPROVED** [review]

File: requirements/requirements.in (hunk 4 above)
Comment (john-westcott-iv) [reply requirements/requirements.in:19]: That is correct. But it should grab the latest compatable version when we run the upgrade script `requirements/updater.sh` and that then produces the requirements.txt file which would have a fixed pinned version:
```
django-oauth-toolkit==1.7.1
```

Comment (jjwatt) [reply requirements/requirements.in:19]: OK. Cool. I thought it might be something like that when I saw the `.in` on it. I'm not in love with that workflow (or that `updater.sh` script since it claims to be `sh` but has bashisms in it), but I'm sure there's a good reason for it that I don't understand yet.
